stan/math/rev/core/autodiffstackstorage.hpp

```diff
@@ -119,7 +119,14 @@ struct AutodiffStackSingleton {
 
  private:
   static bool init() {
+    static STAN_THREADS_DEF bool is_initialized = false;
+    if (!is_initialized) {
+      is_initialized = true;
+      instance_ = new AutodiffStackStorage();
+      return true;
+    }
     if (!instance_) {
+      is_initialized = true;
```

Comment:
Isnt this second if redundant? Is there a case where is_initialized would be true and instance would not exist?

Reply:
Yes, there is: Suppose a thread creates its AD tape instance => is_initialized is true and instance is non-null. Next, the thread destroys for whatever reason the AD tape, but the thread continues to run. Now we have is_initialized=true and instance_=nullptr. At a later stage, the same thread creates again a new tape. At this point we need the second if statement to detect that instance_ is equal to nullptr. The crux with the is_initalized variable is that it can only be modified within the init function... and the good thing is that we know that the variable does exist and can be modified.

Reply:
Ah, that makes total sense. Thanks for the explainer.